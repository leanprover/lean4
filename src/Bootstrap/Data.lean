/-
Copyright (c) 2020 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
import Bootstrap.Data.HashMap
import Bootstrap.Data.HashSet
import Bootstrap.Data.PersistentArray
import Bootstrap.Data.PersistentHashMap
import Bootstrap.Data.PersistentHashSet
import Bootstrap.Data.AssocList
import Bootstrap.Data.RBTree
import Bootstrap.Data.RBMap
