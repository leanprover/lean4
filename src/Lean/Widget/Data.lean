import Lean.Elab.InfoTree
import Lean.Server.Rpc.Basic
import Lean.Widget.TaggedText

/-! This file defines the widget RPC data types but few functions, since in general the UI may explore
them in a mutually recursive fashion:
msg -> trace -> goal -> expr -> info -> msg (only a subset of ctrs) -> ..
-/

namespace Lean.Widget
open Server

-- TODO: Some of the `WithBlah` types exist mostly because we cannot derive multi-argument RPC wrappers.
-- They will be gone eventually.
structure InfoWithCtx where
  ctx : Elab.ContextInfo
  lctx : LocalContext
  info : Elab.Info
  deriving Inhabited, RpcEncoding with { withRef := true }

/-- Pretty-printed syntax (usually but not necessarily an `Expr`) with embedded `Info`s. -/
abbrev CodeWithInfos := TaggedText (WithRpcRef InfoWithCtx)

def CodeWithInfos.pretty (tt : CodeWithInfos) :=
  tt.stripTags

structure InteractiveGoal where
  hyps      : Array (String × CodeWithInfos)
  type      : CodeWithInfos
  userName? : Option String
  deriving Inhabited, RpcEncoding

namespace InteractiveGoal

def pretty (g : InteractiveGoal) : String :=
  let ret := match g.userName? with
    | some userName => s!"case {userName}\n"
    | none          => ""
  let hyps := g.hyps.map fun (name, tt) => s!"{name} : {tt.stripTags}"
  let ret := ret ++ "\n".intercalate hyps.toList
  ret ++ s!"⊢ {g.type.stripTags}"

end InteractiveGoal

-- TODO
structure LazyTrace where
  contents : String
  deriving Inhabited, FromJson, ToJson

inductive MsgEmbed where
  | expr : CodeWithInfos → MsgEmbed
  | goal : InteractiveGoal → MsgEmbed
  | lazyTrace : LazyTrace → MsgEmbed
  deriving Inhabited

namespace MsgEmbed

-- TODO(WN): `deriving RpcEncoding` for `inductive`
private inductive RpcEncodingPacket where
  | expr : TaggedText Lsp.RpcRef → RpcEncodingPacket
  | goal : Nat → RpcEncodingPacket
  | lazyTrace : LazyTrace → RpcEncodingPacket
  deriving Inhabited, FromJson, ToJson

instance : RpcEncoding MsgEmbed RpcEncodingPacket where
  rpcEncode a := match a with
    | expr t      => return RpcEncodingPacket.expr (← rpcEncode t)
    | goal t      => return RpcEncodingPacket.goal 0
    | lazyTrace t => return RpcEncodingPacket.lazyTrace (← rpcEncode t)

  rpcDecode a := match a with
    | RpcEncodingPacket.expr t      => return expr (← rpcDecode t)
    | RpcEncodingPacket.goal t      => return unreachable!
    | RpcEncodingPacket.lazyTrace t => return lazyTrace (← rpcDecode t)

end MsgEmbed

/-- We embed objects in messages by storing them in the tag of an empty subtree (`text ""`).
In other words, we terminate the `MsgEmbed`-tagged tree at embedded objects and instead
store the pretty-printed embed (which can itself be a `TaggedText`) in the tag. -/
abbrev InteractiveMessage := TaggedText MsgEmbed

namespace InteractiveMessage
open MsgEmbed

def pretty (msg : InteractiveMessage) : String :=
  let tt : TaggedText MsgEmbed := msg.rewrite fun
    | expr tt,     _ => TaggedText.text tt.stripTags
    | goal g,      _ => TaggedText.text g.pretty
    | lazyTrace t, _ => TaggedText.text t.contents
  tt.stripTags

end InteractiveMessage

end Lean.Widget
