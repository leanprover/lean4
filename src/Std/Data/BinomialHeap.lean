/-
Copyright (c) 2019 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura, Jannis Limperg
-/
namespace Std
universe u
namespace BinomialHeapImp

structure HeapNodeAux (α : Type u) (h : Type u) where
  val : α
  rank : Nat
  children : List h

-- A `Heap` is a forest of binomial trees.
inductive Heap (α : Type u) : Type u where
  | heap (ns : List (HeapNodeAux α (Heap α))) : Heap α
  deriving Inhabited

open Heap

-- A `HeapNode` is a binomial tree. If a `HeapNode` has rank `k`, its children
-- have ranks between `0` and `k - 1`. They are ordered by rank. Additionally,
-- the value of each child must be less than or equal to the value of its
-- parent node.
abbrev HeapNode α := HeapNodeAux α (Heap α)

variable {α : Type u}

def hRank : List (HeapNode α) → Nat
  | []   => 0
  | h::_ => h.rank

def isEmpty : Heap α → Bool
  | heap [] => true
  | _       => false

def empty : Heap α :=
  heap []

def singleton (a : α) : Heap α :=
  heap [{ val := a, rank := 1, children := [] }]

-- Combine two binomial trees of rank `r`, creating a binomial tree of rank
-- `r + 1`.
@[specialize] def combine (le : α → α → Bool) (n₁ n₂ : HeapNode α) : HeapNode α :=
  if le n₂.val n₁.val then
     { n₂ with rank := n₂.rank + 1, children := n₂.children ++ [heap [n₁]] }
  else
     { n₁ with rank := n₁.rank + 1, children := n₁.children ++ [heap [n₂]] }

-- Merge two forests of binomial trees. The forests are assumed to be ordered
-- by rank and `mergeNodes` maintains this invariant.
@[specialize] partial def mergeNodes (le : α → α → Bool) : List (HeapNode α) → List (HeapNode α) → List (HeapNode α)
  | [], h  => h
  | h,  [] => h
  | f@(h₁ :: t₁), s@(h₂ :: t₂) =>
    if h₁.rank < h₂.rank then h₁ :: mergeNodes le t₁ s
    else if h₂.rank < h₁.rank then h₂ :: mergeNodes le t₂ f
    else
      let merged := combine le h₁ h₂
      let r      := merged.rank
      if r != hRank t₁ then
        if r != hRank t₂ then merged :: mergeNodes le t₁ t₂ else mergeNodes le (merged :: t₁) t₂
      else
        if r != hRank t₂ then mergeNodes le t₁ (merged :: t₂) else merged :: mergeNodes le t₁ t₂

@[specialize] def merge (le : α → α → Bool) : Heap α → Heap α → Heap α
  | heap h₁, heap h₂ => heap (mergeNodes le h₁ h₂)

@[specialize] def head? (le : α → α → Bool) : Heap α → Option α
  | heap []      => none
  | heap (h::hs) => some $
    hs.foldl (init := h.val) fun r n => if le r n.val then r else n.val

@[inline] def head [Inhabited α] (le : α → α → Bool) (h : Heap α) : α :=
  head? le h |>.getD arbitrary

@[specialize] def findMin (le : α → α → Bool) : List (HeapNode α) → Nat → HeapNode α × Nat → HeapNode α × Nat
  | [],    _,   r          => r
  | h::hs, idx, (h', idx') => if le h'.val h.val then findMin le hs (idx+1) (h', idx') else findMin le hs (idx+1) (h, idx)
    -- It is important that we check `le h'.val h.val` here, not the other way
    -- around. This ensures that head? and findMin find the same element even
    -- when we have `le h'.val h.val` and `le h.val h'.val` (i.e. le is not
    -- irreflexive).

def deleteMin (le : α → α → Bool) : Heap α → Option (α × Heap α)
  | heap [] => none
  | heap [h] =>
    let tail :=
      match h.children with
      | [] => empty
      | (h::hs) => hs.foldl (merge le) h
    some (h.val, tail)
  | heap hhs@(h::hs) =>
    let (min, minIdx) := findMin le hs 1 (h, 0)
    let rest          := hhs.eraseIdx minIdx
    let tail          := min.children.foldl (merge le) (heap rest)
    some (min.val, tail)

@[inline] def tail? (le : α → α → Bool) (h : Heap α) : Option (Heap α) :=
  deleteMin le h |>.map (·.snd)

@[inline] def tail (le : α → α → Bool) (h : Heap α) : Heap α :=
  tail? le h |>.getD empty

partial def toList (le : α → α → Bool) (h : Heap α) : List α :=
  match deleteMin le h with
  | none          => []
  | some (hd, tl) => hd :: toList le tl

partial def toArray (le : α → α → Bool) (h : Heap α) : Array α :=
  go #[] h
  where
    go (acc : Array α) (h : Heap α) : Array α :=
      match deleteMin le h with
      | none => acc
      | some (hd, tl) => go (acc.push hd) tl

partial def toListUnordered : Heap α → List α
  | heap ns => ns.bind fun n => n.val :: n.children.bind toListUnordered

partial def toArrayUnordered (h : Heap α) : Array α :=
  go #[] h
  where
    go (acc : Array α) : Heap α → Array α
      | heap ns => do
        let mut acc := acc
        for n in ns do
          acc := acc.push n.val
          for h in n.children do
            acc := go acc h
        return acc

inductive WellFormed (le : α → α → Bool) : Heap α → Prop where
  | empty                : WellFormed le empty
  | singleton (a)        : WellFormed le (singleton a)
  | merge (h₁ h₂)        : WellFormed le h₁ → WellFormed le h₂ → WellFormed le (merge le h₁ h₂)
  | deleteMin (a) (h tl) : WellFormed le h → deleteMin le h = some (a, tl) → WellFormed le tl

theorem WellFormed.tail? {le} (h tl : Heap α) (hwf : WellFormed le h) (eq : tail? le h = some tl) : WellFormed le tl := by
  simp only [BinomialHeapImp.tail?] at eq
  match eq₂: BinomialHeapImp.deleteMin le h with
    | none =>
      rw [eq₂] at eq; cases eq
    | some (a, tl) =>
      rw [eq₂] at eq; cases eq
      exact deleteMin _ _ _ hwf eq₂

theorem WellFormed.tail {le} (h : Heap α) (hwf : WellFormed le h) : WellFormed le (tail le h) := by
  simp only [BinomialHeapImp.tail]
  match eq: BinomialHeapImp.tail? le h with
    | none => exact empty
    | some tl => exact tail? _ _ hwf eq

end BinomialHeapImp

open BinomialHeapImp

def BinomialHeap (α : Type u) (le : α → α → Bool) := { h : Heap α // WellFormed le h }

@[inline] def mkBinomialHeap (α : Type u) (le : α → α → Bool) : BinomialHeap α le :=
  ⟨empty, WellFormed.empty⟩

namespace BinomialHeap
variable {α : Type u} {le : α → α → Bool}

@[inline] def empty : BinomialHeap α le :=
  mkBinomialHeap α le

@[inline] def isEmpty : BinomialHeap α le → Bool
  | ⟨b, _⟩ => BinomialHeapImp.isEmpty b

/- O(1) -/
@[inline] def singleton (a : α) : BinomialHeap α le :=
  ⟨BinomialHeapImp.singleton a, WellFormed.singleton a⟩

/- O(log n) -/
@[inline] def merge : BinomialHeap α le → BinomialHeap α le → BinomialHeap α le
  | ⟨b₁, h₁⟩, ⟨b₂, h₂⟩ => ⟨BinomialHeapImp.merge le b₁ b₂, WellFormed.merge b₁ b₂ h₁ h₂⟩

/- O(log n) -/
@[inline] def insert (a : α) (h : BinomialHeap α le) : BinomialHeap α le :=
  merge (singleton a) h

/- O(n log n) -/
def ofList (le : α → α → Bool) (as : List α) : BinomialHeap α le :=
  as.foldl (flip insert) empty

/- O(n log n) -/
def ofArray (le : α → α → Bool) (as : Array α) : BinomialHeap α le :=
  as.foldl (flip insert) empty

/- O(log n) -/
@[inline] def deleteMin : BinomialHeap α le → Option (α × BinomialHeap α le)
  | ⟨b, h⟩ =>
    match eq: BinomialHeapImp.deleteMin le b with
    | none => none
    | some (a, tl) => some (a, ⟨tl, WellFormed.deleteMin a b tl h eq⟩)

/- O(log n) -/
@[inline] def head [Inhabited α] : BinomialHeap α le → α
  | ⟨b, _⟩ => BinomialHeapImp.head le b

/- O(log n) -/
@[inline] def head? : BinomialHeap α le → Option α
  | ⟨b, _⟩ => BinomialHeapImp.head? le b

/- O(log n) -/
@[inline] def tail? : BinomialHeap α le → Option (BinomialHeap α le)
  | ⟨b, h⟩ =>
    match eq: BinomialHeapImp.tail? le b with
    | none => none
    | some tl => some ⟨tl, WellFormed.tail? b tl h eq⟩

/- O(log n) -/
@[inline] def tail : BinomialHeap α le → BinomialHeap α le
  | ⟨b, h⟩ => ⟨BinomialHeapImp.tail le b, WellFormed.tail b h⟩

/- O(n log n) -/
@[inline] def toList : BinomialHeap α le → List α
  | ⟨b, _⟩ => BinomialHeapImp.toList le b

/- O(n log n) -/
@[inline] def toArray : BinomialHeap α le → Array α
  | ⟨b, _⟩ => BinomialHeapImp.toArray le b

/- O(n) -/
@[inline] def toListUnordered : BinomialHeap α le → List α
  | ⟨b, _⟩ => BinomialHeapImp.toListUnordered b

/- O(n) -/
@[inline] def toArrayUnordered : BinomialHeap α le → Array α
  | ⟨b, _⟩ => BinomialHeapImp.toArrayUnordered b

end BinomialHeap
end Std
