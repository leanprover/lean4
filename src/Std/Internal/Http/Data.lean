/-
Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Sofia Rodrigues
-/
module

prelude
public import Std.Internal.Http.Data.Body
public import Std.Internal.Http.Data.Headers
public import Std.Internal.Http.Data.Method
public import Std.Internal.Http.Data.Version
public import Std.Internal.Http.Data.Request
public import Std.Internal.Http.Data.Response
public import Std.Internal.Http.Data.URI
public import Std.Internal.Http.Data.Status
public import Std.Internal.Http.Data.Version
