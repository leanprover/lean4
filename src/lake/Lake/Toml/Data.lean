/-
Copyright (c) 2024 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
module

prelude
public import Lake.Toml.Data.DateTime
public import Lake.Toml.Data.Dict
public import Lake.Toml.Data.Value

/-!
# TOML Data Types
-/
