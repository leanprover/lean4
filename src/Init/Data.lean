/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Init.Data.Basic
import Init.Data.Nat
import Init.Data.Char
import Init.Data.String
import Init.Data.List
import Init.Data.Int
import Init.Data.Array
import Init.Data.ByteArray
import Init.Data.FloatArray
import Init.Data.Fin
import Init.Data.UInt
import Init.Data.Float
import Init.Data.Option
import Init.Data.Ord
import Init.Data.Random
import Init.Data.ToString
import Init.Data.Range
import Init.Data.Hashable
import Init.Data.OfScientific
import Init.Data.Format
import Init.Data.Stream
