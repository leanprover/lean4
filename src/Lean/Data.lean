/-
Copyright (c) 2020 Sebastian Ullrich. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Sebastian Ullrich
-/
import Lean.Data.AssocList
import Lean.Data.Format
import Lean.Data.HashMap
import Lean.Data.HashSet
import Lean.Data.Json
import Lean.Data.JsonRpc
import Lean.Data.KVMap
import Lean.Data.LBool
import Lean.Data.LOption
import Lean.Data.Lsp
import Lean.Data.Name
import Lean.Data.NameMap
import Lean.Data.OpenDecl
import Lean.Data.Options
import Lean.Data.Parsec
import Lean.Data.PersistentArray
import Lean.Data.PersistentHashMap
import Lean.Data.PersistentHashSet
import Lean.Data.Position
import Lean.Data.PrefixTree
import Lean.Data.SMap
import Lean.Data.Trie
import Lean.Data.Xml
import Lean.Data.NameTrie
import Lean.Data.RBTree
import Lean.Data.RBMap
import Lean.Data.Rat
