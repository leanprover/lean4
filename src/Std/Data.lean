/-
Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Markus Himmel
-/
prelude
import Std.Data.DHashMap
import Std.Data.HashMap
import Std.Data.HashSet
import Std.Data.DTreeMap
import Std.Data.TreeMap
import Std.Data.TreeSet

-- The three imports above only import the modules needed to work with the version which bundles
-- the well-formedness invariant, so we need to additionally import the files that deal with the
-- unbundled version
import Std.Data.DHashMap.RawLemmas
import Std.Data.HashMap.RawLemmas
import Std.Data.HashSet.RawLemmas
