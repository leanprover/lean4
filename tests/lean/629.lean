import Std.Data.BinomialHeap

open Std

def test : BinomialHeap (Nat Ã— Nat) (Î» (x, _) (y, _) => x < y) :=
  BinomialHeap.empty.insert (0, 1) |>.insert (0, 2) |>.insert (0, 3)

#eval test.head
#eval test.tail.toList
