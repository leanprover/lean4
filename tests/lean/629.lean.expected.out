(0, 2)
[(0, 3), (0, 1)]
